pkg/operator/controller/ingress/status.go

```diff
@@ -102,6 +105,73 @@ func computeIngressTLSProfile(oldProfile *configv1.TLSProfileSpec, deployment *a
 	return newProfile
 }
 
+// computeDeploymentPodsScheduledCondition computes the ingress controller's
+// current PodsScheduled status condition state by inspecting the PodScheduled
+// conditions of the pods associated with the deployment.
+func computeDeploymentPodsScheduledCondition(deployment *appsv1.Deployment, pods []corev1.Pod) operatorv1.OperatorCondition {
+	unscheduled := make(map[*corev1.Pod]corev1.PodCondition)
+	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil || selector.Empty() {
+		return operatorv1.OperatorCondition{
+			Type:    IngressControllerPodsScheduledConditionType,
+			Status:  operatorv1.ConditionUnknown,
+			Reason:  "InvalidLabelSelector",
+			Message: "Deployment has an invalid label selector.",
+		}
+	}
+	for i, pod := range pods {
+		if !selector.Matches(labels.Set(pod.Labels)) {
+			continue
+		}
+		for j, cond := range pod.Status.Conditions {
+			if cond.Type != corev1.PodScheduled {
+				continue
+			}
+			if cond.Status == corev1.ConditionTrue {
+				continue
+			}
+			unscheduled[&pods[i]] = pod.Status.Conditions[j]
+		}
+	}
+	if len(unscheduled) != 0 {
+		var haveUnschedulable bool
+		message := "Some pods are not scheduled:"
+		// Sort keys so that the result is deterministic.
```

Comment:
Could you help me understand why sorting `keys` here is necessary?

Reply:
Without sorting, the ordering is nondeterministic.  Suppose you have asked for 2 replicas and have 3 master nodes and 0 worker nodes.  The `PodsSchedulable` status condition reports the following:

      - lastTransitionTime: "2020-09-22T00:51:15Z"
        message: 'Some pods are not scheduled: Pod "router-default-889b9ccdc-5x5z8" cannot be scheduled: 0/3 nodes are available: 3 node(s) didn''t match node selector. Pod "router-default-889b9ccdc-zww95" cannot be scheduled: 0/3 nodes are available: 3 node(s) didn''t match node selector. Make sure you have sufficient worker nodes.'
        reason: PodsNotScheduled
        status: "False"
        type: PodsSchedulable

A moment later, the operator's reconciliation loop runs again and recomputes the status conditions.  This time, the `PodsSchedulable` status condition reports the following:

      - lastTransitionTime: "2020-09-22T00:52:15Z"
        message: 'Some pods are not scheduled: Pod "router-default-889b9ccdc-zww95" cannot be scheduled: 0/3 nodes are available: 3 node(s) didn''t match node selector. Pod "router-default-889b9ccdc-5x5z8" cannot be scheduled: 0/3 nodes are available: 3 node(s) didn''t match node selector. Make sure you have sufficient worker nodes.'
        reason: PodsNotScheduled
        status: "False"
        type: PodsSchedulable

We had bad luck, and the ordering changed.  As a consequence, the message changed, and therefore the status condition's last transition time changed as well.  This means that the last transition time is unreliable: it does not necessarily tell us when the status really changed.

`computeIngressDegradedCondition` scans several status conditions, including (as of this PR) `PodsSchedulable`, to determine whether it should report `Degraded=True`.  Rather than immediately reporting `Degraded=True` the instant it sees something wrong, `computeIngressDegradedCondition` allots each status condition a grace period and checks the status condition's last transition time to determine whether that period has expired.  Therefore `computeIngressDegradedCondition` needs the last transition time to be accurate so that it reports `Degraded=True` once the status condition's grace period has expired; the nondeterministic ordering may prevent it from ever reporting `Degraded=True` when `PodsSchedulable=False`.

Reply:
I see. That makes sense! Thanks for the details.

Reply:
> As a consequence, the message changed, and therefore the status condition's last transition time changed as well.

[API docs][1] say that `lastTransitionTime` is the time of the last `status` change regardless of `reason` or `message` changes.  For example, see [this library-go code][2].

[1]: https://github.com/openshift/api/blob/2a18526802e31a4ea3b0ff6ec807b27f1630074a/config/v1/types_cluster_operator.go#L123
[2]: https://github.com/openshift/library-go/blob/9e6c27e916a0a2aa8eef019683b1c05caecfaa61/pkg/config/clusteroperator/v1helpers/status.go#L29-L31

Reply:
But `lastTransitionTime` aside, I still like constructing the message deterministically to avoid potentially distracting churn on meaningless ordering changes.

Reply:
Isn't this technically a breaking API change?  * grin *

https://github.com/openshift/api/pull/501/commits/40c55f8085093f8dc7b93d70f431011ba74e6fdc#diff-b279e5574ea47c9158c793cb4ea7bf4aL123-R123

We picked up the change in 4.4:

https://github.com/openshift/cluster-ingress-operator/pull/343/commits/5454b6886fdba160253dfe01d459defbe2b8520e#diff-a882102033bf49e063e42db2ed09b51fL123-R123

So this is a long-standing issue, maybe best resolved separately.